cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.messaging.store.hbase;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.data2.util.TableId;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.data2.util.hbase.PutBuilder;
+import co.cask.cdap.messaging.MessagingUtils;
+import co.cask.cdap.messaging.data.MessageId;
+import co.cask.cdap.messaging.store.MessageTable;
+import co.cask.cdap.messaging.store.leveldb.MessageEntry;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.tephra.Transaction;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * HBase implementation of {@link MessageTable}.
+ */
+public class HBaseMessageTable implements MessageTable {
+  private static final byte[] PAYLOAD_COL = Bytes.toBytes('p');
+  private static final byte[] TX_COL = Bytes.toBytes('t');
+  private final Configuration hConf;
+  private final HBaseTableUtil tableUtil;
+  private final TableId tableId;
+  private final byte[] columnFamily;
+  private volatile long writeTimestamp;
+  private volatile short seqId;
+  private volatile HTable hTable;
+
+  public HBaseMessageTable(Configuration hConf, HBaseTableUtil tableUtil, TableId tableId, byte[] columnFamily) {
+    this.hConf = hConf;
+    this.tableUtil = tableUtil;
```

Comment:
Better just have the `HTable` passed in, as there will be some sort of table factory/admin to create instance of this. It simplifies the code. Refer to the `HBaseMetadataTable`.